contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,870 @@
+
+        _hashConsensusForAccountingOracle.grantRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.grantRole(manage_members_role, address(this));
+        for (uint256 i; i < members.length; ++i) {
+            _hashConsensusForAccountingOracle.addMember(members[i], quorum);
+            _hashConsensusForValidatorsExitBusOracle.addMember(members[i], quorum);
+        }
+        _hashConsensusForAccountingOracle.renounceRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.renounceRole(manage_members_role, address(this));
+    }
+
+    function _migrateDSMGuardians() internal {
+        IDepositSecurityModule previousDSM = IDepositSecurityModule(_previousDepositSecurityModule);
+        address[] memory guardians = previousDSM.getGuardians();
+        uint256 quorum = previousDSM.getGuardianQuorum();
+        _depositSecurityModule.addGuardians(guardians, quorum);
+    }
+
+    function _finishUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (!isUpgradeStarted) revert StartMustBeCalledBeforeFinish();
+        if (isUpgradeFinished) revert CanOnlyFinishOnce();
+        /// Here we check that the contract got new ABI function getContractVersion(), although it is 0 yet
+        /// because in the new contract version is stored in a different slot
+        if (_legacyOracle.getContractVersion() != NOT_INITIALIZED_CONTRACT_VERSION) {
+            revert LidoOracleMustBeUpgradedToLegacy();
+        }
+        isUpgradeFinished = true;
+
+        _legacyOracle.finalizeUpgrade_v4(address(_accountingOracle));
+
+        _lido.finalizeUpgrade_v2(address(_locator), _eip712StETH);
+
+        _nodeOperatorsRegistry.finalizeUpgrade_v2(
+            address(_locator),
+            _nodeOperatorsRegistryStakingModuleType,
+            _nodeOperatorsRegistryStuckPenaltyDelay
+        );
+
+        _attachNORToStakingRouter();
+
+        _burner.grantRole(_burner.REQUEST_BURN_SHARES_ROLE(), address(_nodeOperatorsRegistry));
+
+        _passAdminRoleFromTemplateToAgent();
+
+        _assertUpgradeIsFinishedCorrectly();
+    }
+
+    function _attachNORToStakingRouter() internal {
+        bytes32 sm_manage_role = _stakingRouter.STAKING_MODULE_MANAGE_ROLE();
+        _stakingRouter.grantRole(sm_manage_role, address(this));
+        _stakingRouter.addStakingModule(
+            NOR_STAKING_MODULE_NAME,
+            address(_nodeOperatorsRegistry),
+            NOR_STAKING_MODULE_TARGET_SHARE_BP,
+            NOR_STAKING_MODULE_MODULE_FEE_BP,
+            NOR_STAKING_MODULE_TREASURY_FEE_BP
+        );
+        _stakingRouter.renounceRole(sm_manage_role, address(this));
+    }
+
+    function _passAdminRoleFromTemplateToAgent() internal {
+        // NB: No need to pass OracleDaemonConfig and OracleReportSanityChecker admin roles
+        // because they were Agent at the beginning and are not needed by the template
```

Comment:
I really like this comment because it makes the function pretty exhaustive by means of context.  

Suggest to use the same approach here and there across the code for other functions :+1:

Reply:
added